## ■ 판다스 배운 내용 복습

1. 판다스 기본 문법과 판다스 데이터 구조
2. 판다스에서 데이터 검색하는 방법
3. 판다스로 그래프 그리는 방법
    - matplotlib 
    - seaborn

## ■ 6장. 데이터 전처리

1. 누락 데이터 처리
2. 중복 데이터 처리
3. 데이터 표준화
4. 범주형 데이터 처리
5. 정규화
6. 시계열 데이터 생성

### ■ 1. 누락 데이터 처리
   - 누락 데이터 제거 : 열삭제 , 행 삭제
   - 누락 데이터 치환
        1. 평균값
        2. 가장 빈번한 값
        3. 바로 앞행의 값

### ■ 2. 중복 데이터 처리 
   - 중복 데이터를 검색
   - 중복 데이터 제거 하는 방법

### ■ 3. 데이터 표준화
### - 키와 체중처럼 서로 다른 단위가 섞여있거나 같은 대상을 다른 형식으로 표현한 경우 표준화가 필요하다.

   - __3.1 단위환산 :__ 같은 단위로 맞춰주는 작업
   - __3.2 자료형 변환 :__ 숫자가 문자열로 저장된 경우, 숫자(int)로 변환

### ■ 4. 범주형 데이터 처리
   - 4.1 구간 분할
        - 데이터 분석 알고리즘에 따라서 연속데이터를 그대로 사용하기 보다는 일정한 구간으로 나누어서 분석하는것이 효율적인 경우가 있다.  
          가격, 비용, 효율 등 연속적인 값을 일정한 수준이나 정도를  나타내는 이산적인 값으로 나타내어 구간별 차이를 드러내는 것이다.
          
    
          
   - 4.2 더미 변수 생성 : 데이터를 숫자 0 또는 1로 표현하는 변수를 말한다.
        - 카테고리를 나타내는 범주형 데이터는 회귀분석등의 머신러닝 알고리즘에서는 바로 사용할 수 없는 경우가 있는데   
        이럴때는 숫자 0또는 1로 표현되는 더미변수를 사용한다.

### ■ 5. 정규화

데이터 프레임 열에 들어있는 숫자 데이터의 상대적 크기 차이 때문에 머신러닝 분석 결과가 달라질 수 있다.

* A 변수 : 0 ~ 1000 범위
* B 변수 : 0 ~ 1 범위

※ 이 경우는 상대적으로 큰 숫자값을 갖는 A 변수의 영향이 더 커진다.

둘다 min/max 정규화를 써서 0~1 사이의 범위로 만들어 준다.

### ■ 6. 시계열 데이터 생성

일정한 시간 간격으로 배치된 데이터들의 수열을 말한다.

1. 다른 자료형   -------> 시계열 데이터
2. 시계열 데이터 만들기
3. 시계열 데이터 활용

## ■ 6.2 가장 많이 나타나는 데이터로 누락 데이터 바꾸기

__※ 문법__

타이타닉 데이터의 나이 데이터에서 누락 데이터를 뺀 나머지 데이터의 건수중에 가장 건수가 많은 데이터를 most_freq에 담는다.

In [ ]:
# 타이타닉 데이터의 나이 데이터에서 누락 데이터를 뺀 나머지 데이터의 건수중에 가장 건수가 많은 데이터를 most_freq에 담는다.
most_freq = df['age'].value_counts(dropna=True).idxmax() 

# 타이타닉 나이 데이터의 누락 데이터를 most_freq로 변경하겠다.
df['age'].fillna(most_freq, inplace=True)    

### 문제130. seaborn의 타이타닉 데이터에서 embark_town(승선한 도시명)의 데이터의 행을 가져오는데 825행부터 829행까지 검색하시오!

In [3]:
# 라이브러리 불러오기
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd

# titanic 데이터셋 가져오기
df = sns.load_dataset('titanic')

print(df['embark_town'].head(), end= '\n\n')

# 앞부분 열, 뒷부분 행
print(df['embark_town'][825:830], end = '\n\n')

0    Southampton
1      Cherbourg
2    Southampton
3    Southampton
4    Southampton
Name: embark_town, dtype: object

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829            NaN
Name: embark_town, dtype: object



## 문제131. embark_town(승선도시)명을 출력하고 승선한 도시명별 건수를 출력하시오!

In [13]:
import seaborn as sns

df = sns.load_dataset('titanic')

print(df['embark_town'].value_counts(), end = '\n\n\n')

# dropna=False라고 설정하면 누락데이터도 나타나게 해준다.
print(df['embark_town'].value_counts(dropna=False))

Southampton    644
Cherbourg      168
Queenstown      77
Name: embark_town, dtype: int64


Southampton    644
Cherbourg      168
Queenstown      77
NaN              2
Name: embark_town, dtype: int64


## 문제132. 위의 결과중에서 가장 건수가 많은 도시명만 출력하시오!

In [17]:
import seaborn as sns

df = sns.load_dataset('titanic')

# idxmax() 를 활용하여 가장 건수가 많은 도시명만 출력할 수 있다.
most_freq = df['embark_town'].value_counts(dropna=True).idxmax()

print(most_freq)

Southampton


### 문제133. 타이타닉의 승선 도시의 누락 데이터 2건을 가장 빈번하게 나타나는 Southampton으으로 치환하시오.

In [21]:
import seaborn as sns

df = sns.load_dataset('titanic')

# idxmax() 를 활용하여 가장 건수가 많은 도시명만 출력할 수 있다.
most_freq = df['embark_town'].value_counts(dropna=True).idxmax()

df['embark_town'].fillna(most_freq, inplace=True)

print(df['embark_town'][825:830]) # embark_town이 NaN에서 Southampton으로 바뀌었다는것을 확인할 수 있다.

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829    Southampton
Name: embark_town, dtype: object


####  " 타이타닉 생존자 예측을 위한 머신러닝 모델에 학습데이터를 제공할 때 embark_town의 누락 데이터 처리는?? "

1. 가장 빈번한 값으로 치환 하는게 낳을까?
2. 바로 앞행의 값으로 치환 하는게 낳을지?

## 문제134. 이번에는 embark_town의 누락 데이터를 바로 앞행의 값으로 치환시키시오~

In [22]:
import seaborn as sns

df = sns.load_dataset('titanic')

# idxmax() 를 활용하여 가장 건수가 많은 도시명만 출력할 수 있다.
most_freq = df['embark_town'].value_counts(dropna=True).idxmax()

df['embark_town'].fillna(method = 'ffill', inplace=True)

print(df['embark_town'][825:830]) # embark_town이 NaN에서 Southampton으로 바뀌었다는것을 확인할 수 있다.

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829     Queenstown
Name: embark_town, dtype: object


## ■ 6.3 중복 데이터 처리

하나의 데이터셋에서 동일한 관측값이 2개 이상 중복되는 경우 중복 데이터를 찾아서 삭제해야한다.   
왜냐하면 동일한 대상이 중복으로 존재하는 것이므로 분석결과를 왜곡하기 때문입니다.

__문법 :__ df.duplicated() # df 데이터 프레임에 중복된 데이터가 있는지 확인하는 방법.

In [27]:
# 예제

import pandas as pd

df = pd.DataFrame({'c1' : ['a','a','b','a','b'], 'c2' : [1,1,1,2,2], 'c3' : [1,1,2,2,2]})

print(df, end = '\n\n\n')

print(df.duplicated()) # 전에 나온 행들과 비교하여 중복되는 행이면 True로 반환, 처음나온 행이면 False로 반환.

  c1  c2  c3
0  a   1   1
1  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2


0    False
1     True
2    False
3    False
4    False
dtype: bool


### 예제. 위의 c2 컬럼에만 중복된 데이터가 있는지 확인하시오!

In [28]:
# 예제

import pandas as pd

df = pd.DataFrame({'c1' : ['a','a','b','a','b'], 'c2' : [1,1,1,2,2], 'c3' : [1,1,2,2,2]})

print(df, end = '\n\n\n')

print(df['c2'].duplicated()) # 전에 나온 행들과 비교하여 중복되는 행이면 True로 반환, 처음나온 행이면 False로 반환.

# duplicated() 는 True or False 로 결과값을 반환함. (중복데이터 확인용)

  c1  c2  c3
0  a   1   1
1  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2


0    False
1     True
2     True
3    False
4     True
Name: c2, dtype: bool


## 문제135. 아래의 데이터 프레임에서 중복된 행을 제거하시오!

In [33]:
# 예제

import pandas as pd

df = pd.DataFrame({'c1' : ['a','a','b','a','b'], 'c2' : [1,1,1,2,2], 'c3' : [1,1,2,2,2]})

df2 = df.drop_duplicates()
print(df2, end = '\n\n\n')

df3 = df['c2'].drop_duplicates()
print(df3)

# duplicates() 는 중복된 행을 제거하여 DataFrame을 보여줌. (중복 데이터 제거용)

  c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2


0    1
3    2
Name: c2, dtype: int64


## 문제136. seaborn의 타이타닉 데이터의 중복된 데이터가 있는지 확인하시오!

In [40]:
import seaborn as sns

df = sns.load_dataset('titanic')


print(df.duplicated(), end = '\n\n')

print(df.duplicated().sum()) # True를 1로 가정하고 다 더한 값을 나타내줌.

0      False
1      False
2      False
3      False
4      False
       ...  
886     True
887    False
888    False
889    False
890    False
Length: 891, dtype: bool

107


### 생각해야할 문제

'문제136'에서 타이타닉 데이터의 중복된 데이터가 107개가 나오는 이유가 무엇인가?

## ■ 6.4 데이터 표준화

### 1. 단위 환산

실무에서 접하는 데이터셋을 다양한 사람들의 손에 거쳐 만들어진다.  
여러 곳에서 수집한 자료들은 단위 선택, 대소문자 구분, 약칭활용등 여러가지 원인에 의해 다양한 형태로 표현된다.  
잘 정리된 것으로 보이는 자료를 자세히 들어다보면, 서로 다른 단위가 섞여있거나 같은 대상을 다른 형식으로 표현한 경우가 의외로 많다.

__예 :__ 모학생이 모전자에서 모제품의 판매 댓수를 미리 예측하여 생산라인에 다음달에 몇대가 판매가 될거라는것을 통보한다.   
전세계의 대리점에서 엑셀 파일을 보내준다. 이번달 몇대 팔았다는 판매 현황 정보이다.  
__(데이터 전처리에 대한 개념과 코딩 지식이 중요하다!!!)__ 

In [43]:
# 예: UCI 머신러닝 데이터셋 중에 자동차 연비 데이터를 불러옵니다.

import pandas as pd

df = pd.read_csv("d:\\data\\pandas\\auto-mpg.csv")

# 컬럼명 지정하기
df.columns = ['mpg', 'cylinders', 'displacement', 'horsepower', 'weight', 'acceleration', 'model_year', 'origin', 'name']
print(df.head(3))

    mpg  cylinders  displacement horsepower  weight  acceleration  model_year  \
0  15.0          8         350.0      165.0  3693.0          11.5          70   
1  18.0          8         318.0      150.0  3436.0          11.0          70   
2  16.0          8         304.0      150.0  3433.0          12.0          70   

   origin                name  
0       1   buick skylark 320  
1       1  plymouth satellite  
2       1       amc rebel sst  


## ※설명 
mpg열은 mile per gallon의 약자인 컬럼으로 영국과 미국에서는 한국과 달리 갤런당 마일 단위(mile/Gallon)로 연비를 표시합니다.  
한국은 리터당 킬로미터(km/L) 단위로 표시한다.  
따라서 mpg열을 한국에서 사용하는 리터당 킬로미터 단위로 변환하여야 한다.

### 문제137. 1갤런이 3.78541L 1마일이 1.60934km 이다. 그러면 1mpg(mile per gallon)는?

In [44]:
print(1.60934/3.78541) # 0.425 km/L

0.42514285110463595


### 문제138. UCI 자동차 연비를 가지고있는 df 데이터 프레임에 kpl이라는 변수를 추가해서 한국식 연비 단위로 값이 출력되게하시오!

In [46]:
mpg_to_kpl = 1.60934/3.78541
df['kpl'] = df['mpg'] * mpg_to_kpl # 'kpl' 이라는 열을 추가하고 값을 정해줌.
print(df.head(3))

    mpg  cylinders  displacement horsepower  weight  acceleration  model_year  \
0  15.0          8         350.0      165.0  3693.0          11.5          70   
1  18.0          8         318.0      150.0  3436.0          11.0          70   
2  16.0          8         304.0      150.0  3433.0          12.0          70   

   origin                name       kpl  
0       1   buick skylark 320  6.377143  
1       1  plymouth satellite  7.652571  
2       1       amc rebel sst  6.802286  


## 문제139. 위의 kpl 컬럼의 데이터를 소수점이후 2자리까지만 허용되겠금 반올림하시오!

In [48]:
mpg_to_kpl = 1.60934/3.78541

# 'kpl' 이라는 열을 추가하고 값을 정해줌.
# round 함수를 이용하자 !!!
df['kpl'] = round(df['mpg'] * mpg_to_kpl, 2) 

print(df.head(3))

    mpg  cylinders  displacement horsepower  weight  acceleration  model_year  \
0  15.0          8         350.0      165.0  3693.0          11.5          70   
1  18.0          8         318.0      150.0  3436.0          11.0          70   
2  16.0          8         304.0      150.0  3433.0          12.0          70   

   origin                name   kpl  
0       1   buick skylark 320  6.38  
1       1  plymouth satellite  7.65  
2       1       amc rebel sst  6.80  


In [49]:
mpg_to_kpl = 1.60934/3.78541

# 'kpl' 이라는 열을 추가하고 값을 정해줌.
# round 함수를 이용하자 !!!
df['kpl'] = df['mpg'] * mpg_to_kpl
df['kpl'] = df['kpl'].round(2)

print(df.head(3))

    mpg  cylinders  displacement horsepower  weight  acceleration  model_year  \
0  15.0          8         350.0      165.0  3693.0          11.5          70   
1  18.0          8         318.0      150.0  3436.0          11.0          70   
2  16.0          8         304.0      150.0  3433.0          12.0          70   

   origin                name   kpl  
0       1   buick skylark 320  6.38  
1       1  plymouth satellite  7.65  
2       1       amc rebel sst  6.80  


### 2. 데이터 표준화의 자료형 변환

숫자가 문자열로 저장된 경우에 숫자형(int)으로 변환해야 한다.  
숫자 데이터 안에 문자 데이터가 있는지 먼저 검색을 해야한다.  

csv 파일을 데이터 프레임으로 변환하는 과정에서 문자중에 '?'가 섞여있어서 분석이 안되는 경우가 있다.

In [4]:
# 예시
# 모듈 불러오기!
import pandas as pd

# 예: UCI 머신러닝 데이터셋 중에 자동차 연비 데이터를 불러옵니다.
df = pd.read_csv("d:\\data\\pandas\\auto-mpg.csv")
print(df, '\n\n\n')

df.columns = ['mpg', 'cylinders', 'displacement', 'horsepower', 'weight', 'acceleration', 'model_year', 'origin', 'name']

print(df['horsepower'].unique(), '\n') # 'horsepower' 컬럼의 값들의 종류를 확인한다.

print(df)

     18.0  8  307.0  130.0   3504.  12.0  70  1 chevrolet chevelle malibu
0    15.0  8  350.0  165.0  3693.0  11.5  70  1         buick skylark 320
1    18.0  8  318.0  150.0  3436.0  11.0  70  1        plymouth satellite
2    16.0  8  304.0  150.0  3433.0  12.0  70  1             amc rebel sst
3    17.0  8  302.0  140.0  3449.0  10.5  70  1               ford torino
4    15.0  8  429.0  198.0  4341.0  10.0  70  1          ford galaxie 500
..    ... ..    ...    ...     ...   ...  .. ..                       ...
392  27.0  4  140.0  86.00  2790.0  15.6  82  1           ford mustang gl
393  44.0  4   97.0  52.00  2130.0  24.6  82  2                 vw pickup
394  32.0  4  135.0  84.00  2295.0  11.6  82  1             dodge rampage
395  28.0  4  120.0  79.00  2625.0  18.6  82  1               ford ranger
396  31.0  4  119.0  82.00  2720.0  19.4  82  1                chevy s-10

[397 rows x 9 columns] 



['165.0' '150.0' '140.0' '198.0' '220.0' '215.0' '225.0' '190.0' '170.0'
 '160.0' '9

## 문제140. horsepower 열의 문자열 '?'를 NaN값으로 변환하시오!

In [60]:
# 예시
# 모듈 불러오기!
import pandas as pd
import numpy as np

# 예: UCI 머신러닝 데이터셋 중에 자동차 연비 데이터를 불러옵니다.
df = pd.read_csv("d:\\data\\pandas\\auto-mpg.csv")
df.columns = ['mpg', 'cylinders', 'displacement', 'horsepower', 'weight', 'acceleration', 'model_year', 'origin', 'name']

df['horsepower'].replace('?', np.nan, inplace = True) # np.nan으로 바꿔줌 // inplace = True (변경을 적용하겠다.)

print(df['horsepower'].unique())

['165.0' '150.0' '140.0' '198.0' '220.0' '215.0' '225.0' '190.0' '170.0'
 '160.0' '95.00' '97.00' '85.00' '88.00' '46.00' '87.00' '90.00' '113.0'
 '200.0' '210.0' '193.0' nan '100.0' '105.0' '175.0' '153.0' '180.0'
 '110.0' '72.00' '86.00' '70.00' '76.00' '65.00' '69.00' '60.00' '80.00'
 '54.00' '208.0' '155.0' '130.0' '112.0' '92.00' '145.0' '137.0' '158.0'
 '167.0' '94.00' '107.0' '230.0' '49.00' '75.00' '91.00' '122.0' '67.00'
 '83.00' '78.00' '52.00' '61.00' '93.00' '148.0' '129.0' '96.00' '71.00'
 '98.00' '115.0' '53.00' '81.00' '79.00' '120.0' '152.0' '102.0' '108.0'
 '68.00' '58.00' '149.0' '89.00' '63.00' '48.00' '66.00' '139.0' '103.0'
 '125.0' '133.0' '138.0' '135.0' '142.0' '77.00' '62.00' '132.0' '84.00'
 '64.00' '74.00' '116.0' '82.00']


## 문제141. dropna(axis = 0) 메소드로 NaN값이 들어있는 모든행을 삭제하시오.

In [6]:
# 예시
# 모듈 불러오기!
import pandas as pd
import numpy as np

# 예: UCI 머신러닝 데이터셋 중에 자동차 연비 데이터를 불러옵니다.
df = pd.read_csv("d:\\data\\pandas\\auto-mpg.csv")
df.columns = ['mpg', 'cylinders', 'displacement', 'horsepower', 'weight', 'acceleration', 'model_year', 'origin', 'name']

df['horsepower'].replace('?', np.nan, inplace = True) # np.nan으로 바꿔줌 // inplace = True (변경을 적용하겠다.)

df.dropna(subset = ['horsepower'], axis = 0, inplace = True) # NaN =누락데이터를 제거함.

print(df['horsepower'].dtypes)

print(df['horsepower'].isnull().sum())

object
0


## 문제142. 위의 horsepower의 데이터 유형을 실수형인 float으로 변환하시오! (점심시간문제)


df['열이름'].astype('데이터 유형타입')

In [72]:
# 예시
# 모듈 불러오기!
import pandas as pd
import numpy as np

# 예: UCI 머신러닝 데이터셋 중에 자동차 연비 데이터를 불러옵니다.
df = pd.read_csv("d:\\data\\pandas\\auto-mpg.csv")
df.columns = ['mpg', 'cylinders', 'displacement', 'horsepower', 'weight', 'acceleration', 'model_year', 'origin', 'name']

df['horsepower'].replace('?', np.nan, inplace = True) # np.nan으로 바꿔줌 // inplace = True (변경을 적용하겠다.)

df.dropna(subset = ['horsepower'], axis = 0, inplace = True)

df['horsepower'] = df['horsepower'].astype('float')

print(df['horsepower'].unique())

print(df['horsepower'].dtypes)

[165. 150. 140. 198. 220. 215. 225. 190. 170. 160.  95.  97.  85.  88.
  46.  87.  90. 113. 200. 210. 193. 100. 105. 175. 153. 180. 110.  72.
  86.  70.  76.  65.  69.  60.  80.  54. 208. 155. 130. 112.  92. 145.
 137. 158. 167.  94. 107. 230.  49.  75.  91. 122.  67.  83.  78.  52.
  61.  93. 148. 129.  96.  71.  98. 115.  53.  81.  79. 120. 152. 102.
 108.  68.  58. 149.  89.  63.  48.  66. 139. 103. 125. 133. 138. 135.
 142.  77.  62. 132.  84.  64.  74. 116.  82.]
float64


## ■ 6.5 범주형 데이터 처리

   __1. 구간 분할 :__ 연속형 데이터 ---> 일정한 구간으로 나눔  
       예) 사원 테이블의 월급의 경우 고소득자, 중간소득자, 저소득자로 나눈다.
       
       
   __2. 더미 변수 :__ 숫자 0또는 1로 표현하는 것  
       예) UCI 연비 데이터의 MPG를 고출력, 저출력으로 변경했다면, 고출력을 1로 저출력을 0으로 변경

### ■ 구간 분할 코드

In [92]:
# 예시
# 모듈 불러오기!
import pandas as pd
import numpy as np

# 예: UCI 머신러닝 데이터셋 중에 자동차 연비 데이터를 불러옵니다.
df = pd.read_csv("d:\\data\\pandas\\auto-mpg.csv")
df.columns = ['mpg', 'cylinders', 'displacement', 'horsepower', 'weight', 'acceleration', 'model_year', 'origin', 'name']

df['horsepower'].replace('?', np.nan, inplace = True) # np.nan으로 바꿔줌 // inplace = True (변경을 적용하겠다.)
df.dropna(subset = ['horsepower'], axis = 0, inplace = True) # 누락 데이터를 제거 
df['horsepower'] = df['horsepower'].astype('float')

count, bin_dividers = np.histogram(df['horsepower'], bins=3)
print(count) # 각 구간에 속하는 값의 갯수 [257 102  32] 
print(bin_dividers) # 경계값 리스트 [ 46.         107.33333333 168.66666667 230.  ]

bin_names = ['저출력', '보통출력', '고출력']
df['hp_bin'] = pd.cut(x = df['horsepower'], bins = bin_dividers, labels = bin_names, include_lowest = True) # 첫 경계값 포함

# 데이터 배열: x = df['horsepower']
# 경계값 리스트: bins = bin_dividers
# bin 이름: labels = bin_names
# 첫 경계값 포함 유무: include_lowest = True

print(df[['horsepower', 'hp_bin']].head(15))

[257 102  32]
[ 46.         107.33333333 168.66666667 230.        ]
    horsepower hp_bin
0        165.0   보통출력
1        150.0   보통출력
2        150.0   보통출력
3        140.0   보통출력
4        198.0    고출력
5        220.0    고출력
6        215.0    고출력
7        225.0    고출력
8        190.0    고출력
9        170.0    고출력
10       160.0   보통출력
11       150.0   보통출력
12       225.0    고출력
13        95.0    저출력
14        95.0    저출력


### 설명:

이처럼 연속 변수를 일정한 구간으로 나누고, 각 구간을 범주형 이산변수로 변환하는 과정을 __구간분할(binning)__이라고 합니다.  
판다스 cut() __함수(pd.cut)__를 이용하면 연속 데이터를 여러구간으로 나누고 범주형 데이터로 변환 할 수 있습니다.  

## 문제143. emp.csv를 판다스로 불러와서 이름, 월급, 월급을 구간 분할한 결과를 출력하시오!
## (고소득, 중간소득, 저소득)

In [132]:
import numpy as np
import pandas as pd

emp = pd.read_csv("d:\\data\\emp.csv", header = 0)
print(emp.columns, end = '\n\n')

emp['sal'] = emp['sal'].astype("float")
count, bin_dividers = np.histogram(emp['sal'], bins=3)

bin_names = ['저소득', ' 중간소득', '고소득']

emp['sal_bin'] = pd.cut(x = emp['sal'], bins = bin_dividers, labels = bin_names, include_lowest = True)

print(emp[['ename', 'sal', 'sal_bin']].head(15))

Index(['empno', 'ename', 'job', 'mgr', 'hiredate', 'sal', 'comm', 'deptno'], dtype='object')

     ename     sal sal_bin
0     KING  5000.0     고소득
1    BLAKE  2850.0    중간소득
2    CLARK  2450.0    중간소득
3    JONES  2975.0    중간소득
4   MARTIN  1250.0     저소득
5    ALLEN  1600.0     저소득
6   TURNER  1500.0     저소득
7    JAMES   950.0     저소득
8     WARD  1250.0     저소득
9     FORD  3000.0    중간소득
10   SMITH   800.0     저소득
11   SCOTT  3000.0    중간소득
12   ADAMS  1100.0     저소득
13  MILLER  1300.0     저소득


## ■ 더미 변수 

머신러닝 알고리즘에 따라 범주형 데이터를 바로 사용할 수 없는 알고리즘이 있는데 이럴때는 컴퓨터가 인식 가능한 입력값으로 변환을 해주어야 
한다.  
이럴때 __숫자 0 또는 1로 표현__되는 __더미변수(dummy variable)__를 사용합니다.  
__여기서 0과 1은__ 수의 크고 작음을 나타내지는 않고, __어떤 특성이 있는지 없는지 여부만 표시한다,__

In [ ]:
horsepower_dumies = pd.get_dummies(df['hp_bin']) # 더미 특성 추가하는 코드
print(horsepower_dummies.head(15)) # 더미 특성 추가되었는지 확인!

In [7]:
# 예시
# 모듈 불러오기!
import pandas as pd
import numpy as np

# 예: UCI 머신러닝 데이터셋 중에 자동차 연비 데이터를 불러옵니다.
df = pd.read_csv("d:\\data\\pandas\\auto-mpg.csv")
df.columns = ['mpg', 'cylinders', 'displacement', 'horsepower', 'weight', 'acceleration', 'model_year', 'origin', 'name']

df['horsepower'].replace('?', np.nan, inplace = True) # np.nan으로 바꿔줌 // inplace = True (변경을 적용하겠다.)
df.dropna(subset = ['horsepower'], axis = 0, inplace = True) # 누락 데이터를 제거 
df['horsepower'] = df['horsepower'].astype('float')

count, bin_dividers = np.histogram(df['horsepower'], bins=3)
print(count) # 각 구간에 속하는 값의 갯수 [257 102  32] 
print(bin_dividers) # 경계값 리스트 [ 46.         107.33333333 168.66666667 230.  ]

bin_names = ['저출력', '보통출력', '고출력']
df['hp_bin'] = pd.cut(x = df['horsepower'], bins = bin_dividers, labels = bin_names, include_lowest = True) # 첫 경계값 포함유무
# 데이터 배열: x = df['horsepower']
# 경계값 리스트: bins = bin_dividers
# bin 이름: labels = bin_names
# 첫 경계값 포함 유무: include_lowest = True

print(df[['horsepower', 'hp_bin']].head(15), '\n\n\n')

horsepower_dumies = pd.get_dummies( df['hp_bin'] )

print ( horsepower_dumies.head(15)  )

[257 102  32]
[ 46.         107.33333333 168.66666667 230.        ]
    horsepower hp_bin
0        165.0   보통출력
1        150.0   보통출력
2        150.0   보통출력
3        140.0   보통출력
4        198.0    고출력
5        220.0    고출력
6        215.0    고출력
7        225.0    고출력
8        190.0    고출력
9        170.0    고출력
10       160.0   보통출력
11       150.0   보통출력
12       225.0    고출력
13        95.0    저출력
14        95.0    저출력 



hp_bin  저출력  보통출력  고출력
0         0     1    0
1         0     1    0
2         0     1    0
3         0     1    0
4         0     0    1
5         0     0    1
6         0     0    1
7         0     0    1
8         0     0    1
9         0     0    1
10        0     1    0
11        0     1    0
12        0     0    1
13        1     0    0
14        1     0    0


## 문제144. 사원 테이블의 월급을 더미변수로 생성하시오!

In [9]:
import numpy as np
import pandas as pd

emp = pd.read_csv("d:\\data\\emp.csv", header = 0)
print(emp.columns)

emp['sal'] = emp['sal'].astype("float")
count, bin_dividers = np.histogram(emp['sal'], bins=3)

bin_names = ['저소득', ' 중간소득', '고소득']

emp['sal_bin'] = pd.cut(x = emp['sal'], bins = bin_dividers, labels = bin_names, include_lowest = True) 
# 데이터 배열: x = emp['sal']
# 경계값 리스트: bins = bin_dividers
# bin 이름: labels = bin_names
# 첫 경계값 포함 유무: include_lowest = True

sal_dumies = pd.get_dummies( emp['sal_bin'] )

print(sal_dumies.head(15))

Index(['empno', 'ename', 'job', 'mgr', 'hiredate', 'sal', 'comm', 'deptno'], dtype='object')
sal_bin  저소득   중간소득  고소득
0          0      0    1
1          0      1    0
2          0      1    0
3          0      1    0
4          1      0    0
5          1      0    0
6          1      0    0
7          1      0    0
8          1      0    0
9          0      1    0
10         1      0    0
11         0      1    0
12         1      0    0
13         1      0    0


## 문제145. 위의 결과에 맨 앞에 이름을 출력하시오!

In [6]:
import numpy as np
import pandas as pd

emp = pd.read_csv("d:\\data\\emp.csv", header = 0)

emp['sal'] = emp['sal'].astype("float")
count, bin_dividers = np.histogram(emp['sal'], bins=3)

bin_names = ['저소득', ' 중간소득', '고소득']

emp['sal_bin'] = pd.cut(x = emp['sal'], bins = bin_dividers, labels = bin_names, include_lowest = True) 
# 데이터 배열: x = emp['sal']
# 경계값 리스트: bins = bin_dividers
# bin 이름: labels = bin_names
# 첫 경계값 포함 유무: include_lowest = True

sal_dumies = pd.get_dummies( emp['sal_bin'] ) # 더미 변수 만들기 코드

print(emp[['ename']].join(sal_dumies.head(10)), end = '\n\n\n')
print(emp[['ename']].join(sal_dumies)) 
print('\n')

print(emp)

     ename  저소득   중간소득  고소득
0     KING  0.0    0.0  1.0
1    BLAKE  0.0    1.0  0.0
2    CLARK  0.0    1.0  0.0
3    JONES  0.0    1.0  0.0
4   MARTIN  1.0    0.0  0.0
5    ALLEN  1.0    0.0  0.0
6   TURNER  1.0    0.0  0.0
7    JAMES  1.0    0.0  0.0
8     WARD  1.0    0.0  0.0
9     FORD  0.0    1.0  0.0
10   SMITH  NaN    NaN  NaN
11   SCOTT  NaN    NaN  NaN
12   ADAMS  NaN    NaN  NaN
13  MILLER  NaN    NaN  NaN


     ename  저소득   중간소득  고소득
0     KING    0      0    1
1    BLAKE    0      1    0
2    CLARK    0      1    0
3    JONES    0      1    0
4   MARTIN    1      0    0
5    ALLEN    1      0    0
6   TURNER    1      0    0
7    JAMES    1      0    0
8     WARD    1      0    0
9     FORD    0      1    0
10   SMITH    1      0    0
11   SCOTT    0      1    0
12   ADAMS    1      0    0
13  MILLER    1      0    0


    empno   ename        job     mgr    hiredate     sal    comm  deptno  \
0    7839    KING  PRESIDENT     NaN  1981-11-17  5000.0     NaN      10   
1   

## ■ 더미변수를 생성하는 다른 방법

sklearn 라이브러리를 이용해서 __원핫인코딩__을 편하게 처리할 수 있다.  
데이터 프레임 df의 'hp_bin'열에 들어있는 범주형 데이터를 0과 1로 __원핫인코딩__ 해서 벡터로 반환할 수 있습니다.

In [121]:
from sklearn import preprocessing

label_encoder = preprocessing.LabelEncoder()
onehot_encoder = preprocessing.OneHotEncoder()

onehot_labeled = label_encoder.fit_transform(df['hp_bin'].head(15))
print(onehot_labeled)

[1 1 1 1 0 0 0 0 0 0 1 1 0 2 2]


In [4]:
import pandas as pd
import numpy as np
from sklearn import preprocessing

# 예: UCI 머신러닝 데이터셋 중에 자동차 연비 데이터를 불러옵니다.
df = pd.read_csv("d:\\data\\pandas\\auto-mpg.csv")
df.columns = ['mpg', 'cylinders', 'displacement', 'horsepower', 'weight', 'acceleration', 'model_year', 'origin', 'name']

df['horsepower'].replace('?', np.nan, inplace = True) # np.nan으로 바꿔줌 // inplace = True (변경을 적용하겠다.)
df.dropna(subset = ['horsepower'], axis = 0, inplace = True) # 누락 데이터를 제거 
df['horsepower'] = df['horsepower'].astype('float')

count, bin_dividers = np.histogram(df['horsepower'], bins=3)
print(count) # 각 구간에 속하는 값의 갯수 [257 102  32] 
print(bin_dividers, end = '\n\n\n') # 경계값 리스트 [ 46.         107.33333333 168.66666667 230.  ]

bin_names = ['저출력', '보통출력', '고출력']
df['hp_bin'] = pd.cut(x = df['horsepower'], bins = bin_dividers, labels = bin_names, include_lowest = True) # 첫 경계값 포함유무

label_encoder = preprocessing.LabelEncoder()
onehot_encoder = preprocessing.OneHotEncoder()

onehot_labeled = label_encoder.fit_transform(df['hp_bin'].head(15))
# 1차원 벡터를 2차원 행렬로 변환한다.
onehot_reshaped = onehot_labeled.reshape(len(onehot_labeled), 1)

# 2차원 행렬을 희소행렬로 변환한다. 희소행렬은 '(행,열) 좌표'와 '값'의 형태로 정리된다.
onehot_encode = onehot_encoder.fit_transform(onehot_reshaped)

print(onehot_labeled)
print(onehot_reshaped)
print(onehot_encode)

[257 102  32]
[ 46.         107.33333333 168.66666667 230.        ]


[1 1 1 1 0 0 0 0 0 0 1 1 0 2 2]
[[1]
 [1]
 [1]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [1]
 [0]
 [2]
 [2]]
  (0, 1)	1.0
  (1, 1)	1.0
  (2, 1)	1.0
  (3, 1)	1.0
  (4, 0)	1.0
  (5, 0)	1.0
  (6, 0)	1.0
  (7, 0)	1.0
  (8, 0)	1.0
  (9, 0)	1.0
  (10, 1)	1.0
  (11, 1)	1.0
  (12, 0)	1.0
  (13, 2)	1.0
  (14, 2)	1.0


C:\Users\wdp\Anaconda3\lib\site-packages\sklearn\preprocessing\_encoders.py:415: FutureWarning: The handling of integer data will change in version 0.22. Currently, the categories are determined based on the range [0, max(values)], while in the future they will be determined based on the unique values.
If you want the future behaviour and silence this warning, you can specify "categories='auto'".
In case you used a LabelEncoder before this OneHotEncoder to convert the categories to integers, then you can now use the OneHotEncoder directly.
  warnings.warn(msg, FutureWarning)


###  ※코딩 결과값 설명
(0,1)은 0행의 1열의 위치를 말하고 데이터 값은 숫자 1.0이 입력된다.

## 6.6 정규화

각 변수(데이터프레임의 열)에 들어있는 숫자 데이터의 상대적 크기 차이 때문에 머신러닝 분석 결과가 달라질 수 있다.  
따라서 숫자 데이터의 상대적인 크기 차이를 제거할 필요가 있다.  
각 열(변수)에 속하는 데이터 값을 동일한 크기 기준으로 나눈 비율로 나타내는 것을 __정규화(normalization)__라고 합니다.  
정규화 과정을 거친 데이터의 범위는 0~1 또는 -1 ~ 1 이 됩니다.

### 문제146. UCI 연비 데이터 프레임의 horsepower 데이터에서 '?'를 결측치로 대체하고 결측치를 제거하여 float로 변환한 데이터를 출력하시오!

In [135]:
# 예시
# 모듈 불러오기!
import pandas as pd
import numpy as np

# 예: UCI 머신러닝 데이터셋 중에 자동차 연비 데이터를 불러옵니다.
df = pd.read_csv("d:\\data\\pandas\\auto-mpg.csv")
df.columns = ['mpg', 'cylinders', 'displacement', 'horsepower', 'weight', 'acceleration', 'model_year', 'origin', 'name']

df['horsepower'].replace('?', np.nan, inplace = True) # np.nan으로 바꿔줌 // inplace = True (변경을 적용하겠다.)
df.dropna(subset = ['horsepower'], axis = 0, inplace = True)
df['horsepower'] = df['horsepower'].astype('float')

print(df['horsepower'].unique())

[165. 150. 140. 198. 220. 215. 225. 190. 170. 160.  95.  97.  85.  88.
  46.  87.  90. 113. 200. 210. 193. 100. 105. 175. 153. 180. 110.  72.
  86.  70.  76.  65.  69.  60.  80.  54. 208. 155. 130. 112.  92. 145.
 137. 158. 167.  94. 107. 230.  49.  75.  91. 122.  67.  83.  78.  52.
  61.  93. 148. 129.  96.  71.  98. 115.  53.  81.  79. 120. 152. 102.
 108.  68.  58. 149.  89.  63.  48.  66. 139. 103. 125. 133. 138. 135.
 142.  77.  62. 132.  84.  64.  74. 116.  82.]


## 문제147. 위의 horsepower 데이터의 통계요약을 확인하시오!( describe()를 사용하시오 ) 

In [138]:
# 예시
# 모듈 불러오기!
import pandas as pd
import numpy as np

# 예: UCI 머신러닝 데이터셋 중에 자동차 연비 데이터를 불러옵니다.
df = pd.read_csv("d:\\data\\pandas\\auto-mpg.csv")
df.columns = ['mpg', 'cylinders', 'displacement', 'horsepower', 'weight', 'acceleration', 'model_year', 'origin', 'name']

df['horsepower'].replace('?', np.nan, inplace = True) # np.nan으로 바꿔줌 // inplace = True (변경을 적용하겠다.)
df.dropna(subset = ['horsepower'], axis = 0, inplace = True)
df['horsepower'] = df['horsepower'].astype('float')

print(df['horsepower'].describe(), end= '\n\n')
print(df.horsepower.describe())

count    391.000000
mean     104.404092
std       38.518732
min       46.000000
25%       75.000000
50%       93.000000
75%      125.000000
max      230.000000
Name: horsepower, dtype: float64

count    391.000000
mean     104.404092
std       38.518732
min       46.000000
25%       75.000000
50%       93.000000
75%      125.000000
max      230.000000
Name: horsepower, dtype: float64


## 문제148. 위의 horsepower 데이터를 0~1 사이의 데이터로 정규화하시오!
## (각 값들의 가장 큰값으로 나누면 됩니다.)

In [147]:
# 예시
# 모듈 불러오기!
import pandas as pd
import numpy as np

# 예: UCI 머신러닝 데이터셋 중에 자동차 연비 데이터를 불러옵니다.
df = pd.read_csv("d:\\data\\pandas\\auto-mpg.csv")
df.columns = ['mpg', 'cylinders', 'displacement', 'horsepower', 'weight', 'acceleration', 'model_year', 'origin', 'name']

df['horsepower'].replace('?', np.nan, inplace = True) # np.nan으로 바꿔줌 // inplace = True (변경을 적용하겠다.)
df.dropna(subset = ['horsepower'], axis = 0, inplace = True)
df['horsepower'] = df['horsepower'].astype('float')

df['horsepower'] = df['horsepower'] / abs(df['horsepower'].max())

print(df['horsepower'].head(), end = '\n\n')
print(df['horsepower'].describe())

0    0.717391
1    0.652174
2    0.652174
3    0.608696
4    0.860870
Name: horsepower, dtype: float64

count    391.000000
mean       0.453931
std        0.167473
min        0.200000
25%        0.326087
50%        0.404348
75%        0.543478
max        1.000000
Name: horsepower, dtype: float64


### 문제149.  각 열(변수)의 데이터 중에서 최대값(max)와 최소값(min)을 뺀 값으로 나누는 방법을 사용하시오. 각 열 데이터에서 해당 열의 최소값을 뺀 값을 분자로 하고, 해당 열의 최대값과 최소값의 차를 분모로 하는 수를 계산하면 가장 큰 값은 역시 1이 됩니다.

In [151]:
# 두번째 정규화 방법
# 모듈 불러오기!
import pandas as pd
import numpy as np

# 예: UCI 머신러닝 데이터셋 중에 자동차 연비 데이터를 불러옵니다.
df = pd.read_csv("d:\\data\\pandas\\auto-mpg.csv")
df.columns = ['mpg', 'cylinders', 'displacement', 'horsepower', 'weight', 'acceleration', 'model_year', 'origin', 'name']

df['horsepower'].replace('?', np.nan, inplace = True) # np.nan으로 바꿔줌 // inplace = True (변경을 적용하겠다.)
df.dropna(subset = ['horsepower'], axis = 0, inplace = True)
df['horsepower'] = df['horsepower'].astype('float')

min_x = df.horsepower - df.horsepower.min() 
min_max = df.horsepower.max() - df.horsepower.min() # df['horsepower'].max() - df['horsepower'].min()
df.horsepower = min_x / min_max

print(df.horsepower.head(), end = '\n\n')
print(df.horsepower.describe())

0    0.646739
1    0.565217
2    0.565217
3    0.510870
4    0.826087
Name: horsepower, dtype: float64

count    391.000000
mean       0.317414
std        0.209341
min        0.000000
25%        0.157609
50%        0.255435
75%        0.429348
max        1.000000
Name: horsepower, dtype: float64


## ■ 6.7 판다스로 시계열 데이터 생성하기

### 일정한 시간 간격으로 배치된 데이터들의 수열을 말한다.

__다른자료형 ---------------------->  시계열 데이터로 변환__

주식, 환율등의 금융 데이터를 다루기 위해 개발된 판다스는 시계열(time series) 데이터를 다루는 여러가지 유용한 기능을 제공한다.    
특히, 시계열 데이터를 데이터프레임의 행 인덱스로 사용하면 시간으로 기록된 데이터를 분석하는것이 매우 편리합니다.

### 1. 문자열을 to_datetime() 함수를 이용하여 timestamp로 변환하기

In [156]:
import pandas as pd

df = pd.read_csv("d:\\data\\pandas\\stock-data.csv")
print(df.head(), end = '\n\n')
print(df.info()) # Date 컬럼의 자료형이 무엇이지 알수 있음.

         Date  Close  Start   High    Low  Volume
0  2018-07-02  10100  10850  10900  10000  137977
1  2018-06-29  10700  10550  10900   9990  170253
2  2018-06-28  10400  10900  10950  10150  155769
3  2018-06-27  10900  10800  11050  10500  133548
4  2018-06-26  10800  10900  11000  10700   63039

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 6 columns):
Date      20 non-null object
Close     20 non-null int64
Start     20 non-null int64
High      20 non-null int64
Low       20 non-null int64
Volume    20 non-null int64
dtypes: int64(5), object(1)
memory usage: 1.1+ KB
None


### ※ info() 메소드로 Date 열의 자료형을 확인해보니 문자열(object)임을 알 수 있다.  
### 판다스의 to_datetime 함수를 써서 Timestamp 객체로 변경한다.

In [7]:
import pandas as pd

df = pd.read_csv("d:\\data\\pandas\\stock-data.csv")

df['new_Date'] = pd.to_datetime(df['Date']) # 새로운 컬럼을 추가
print(df, end = '\n\n')
print(df.info())

          Date  Close  Start   High    Low  Volume   new_Date
0   2018-07-02  10100  10850  10900  10000  137977 2018-07-02
1   2018-06-29  10700  10550  10900   9990  170253 2018-06-29
2   2018-06-28  10400  10900  10950  10150  155769 2018-06-28
3   2018-06-27  10900  10800  11050  10500  133548 2018-06-27
4   2018-06-26  10800  10900  11000  10700   63039 2018-06-26
5   2018-06-25  11150  11400  11450  11000   55519 2018-06-25
6   2018-06-22  11300  11250  11450  10750  134805 2018-06-22
7   2018-06-21  11200  11350  11750  11200  133002 2018-06-21
8   2018-06-20  11550  11200  11600  10900  308596 2018-06-20
9   2018-06-19  11300  11850  11950  11300  180656 2018-06-19
10  2018-06-18  12000  13400  13400  12000  309787 2018-06-18
11  2018-06-15  13400  13600  13600  12900  201376 2018-06-15
12  2018-06-14  13450  13200  13700  13150  347451 2018-06-14
13  2018-06-12  13200  12200  13300  12050  558148 2018-06-12
14  2018-06-11  11950  12000  12250  11950   62293 2018-06-11
15  2018

__※설명 :__ 위와 같이 datetime으로 변경이 되면 시간 순서에 맞춰 데이터를 검색하기가 편리해진다. 시계열 컬럼을 로우 인덱스로 지정하면  
시간 순서에 맞춰 인덱싱 또는 슬라이싱 하기가 편해진다.

## 문제150. 지금 추가한 new_Date 컬럼을 행 인덱스로 지정하고 기존의 Date 컬럼을 삭제하시오!

In [159]:
import pandas as pd

df = pd.read_csv("d:\\data\\pandas\\stock-data.csv")

df['new_Date'] = pd.to_datetime(df['Date']) # 새로운 컬럼을 추가
df.set_index('new_Date', inplace = True)
df.drop('Date', axis = 1, inplace = True) # axis = 1 (열)

print(df.head(), end = '\n\n')
print(df.info())

            Close  Start   High    Low  Volume
new_Date                                      
2018-07-02  10100  10850  10900  10000  137977
2018-06-29  10700  10550  10900   9990  170253
2018-06-28  10400  10900  10950  10150  155769
2018-06-27  10900  10800  11050  10500  133548
2018-06-26  10800  10900  11000  10700   63039

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 20 entries, 2018-07-02 to 2018-06-01
Data columns (total 5 columns):
Close     20 non-null int64
Start     20 non-null int64
High      20 non-null int64
Low       20 non-null int64
Volume    20 non-null int64
dtypes: int64(5)
memory usage: 960.0 bytes
None


### 문제151. 위의 new_Date 시계열 데이터를 이용하여 새로운 파생변수 3가지를 생성하시오! ( Year, Month, Day)
### Year 변수에는 2018과 같이 년도만 담고
### Month 변수에는 7,8 처럼 월 데이터만 담으시오!
### Day 변수에는 29,28 처럼 일 데이터만 담으시오!

In [162]:
import pandas as pd
df = pd.read_csv("d:\\data\\pandas\\stock-data.csv")

df['new_Date'] = pd.to_datetime(df['Date']) # 새로운 컬럼을 추가

df['Year'] = df['new_Date'].dt.year
df['Month'] = df['new_Date'].dt.month
df['Day'] = df['new_Date'].dt.day

print(df.head())

         Date  Close  Start   High    Low  Volume   new_Date  Year  Month  Day
0  2018-07-02  10100  10850  10900  10000  137977 2018-07-02  2018      7    2
1  2018-06-29  10700  10550  10900   9990  170253 2018-06-29  2018      6   29
2  2018-06-28  10400  10900  10950  10150  155769 2018-06-28  2018      6   28
3  2018-06-27  10900  10800  11050  10500  133548 2018-06-27  2018      6   27
4  2018-06-26  10800  10900  11000  10700   63039 2018-06-26  2018      6   26


## ■ 7장. 판다스 데이터 프레임의 응용

   - 7.1 개별 원소에 함수 매핑
       - 시리즈 원소에 함수 매핑( 문법 : 시리즈 객체.apply('매핑함수'))
       - 데이터 프레임 원소에 함수 매핑( 문법 : 데이터프레임객체.applymap('매핑함수'))
       
       
   - 7.2 시리즈 객체에 함수 매핑 (axis 부분 잘 주시해서 외워두기)
       - 데이터 프레임의 각 열에 함수 매핑( 문법 : 데이터프레임객체.apply('매핑함수', axis = 0))
       - 데이터 프레임의 각 행에 함수 매핑( 문법 : 데이터프레임객체.apply('매핑함수', axis = 1))
       - 데이터 프레임 객체의 함수 매핑( 문법: 데이터프레임객체.pipe('매핑함수')
       
       
   - 7.3 열의 재구성
       - 열순서 변경( 문법 : 데이터프레임 객체([재구성할 열이름 리스트])
       - 열분리 : 하나의 열이 여러가지 정보를 담고 있을 때 각 정보를 서로 분리해서 사용하는 경우가 많다.
       
       
   - 7.4 필터링 : 시리즈 또는 데이터 프레임의 데이터 중에서 특정 조건에 만족하는 원소만 따로 추출하는 개념
       - 불린 인덱싱
       - isin 메소드
       
       
   - 7.5 데이터 프레임 합치기 : 데이터가 여러 군데 나누어져 있을때 하나로 합치거나 데이터를 연결해야하는 경우가 있다.
       - 데이터 프레임 연결 ( 문법 : pandas.concat(데이터프레임리스트) )
       - 데이터 프레임 병합 ( 문법 : pandas.merge(emp,dept, how = 'inner') 
       - 데이터 프레임 결합 ( 문법 : DataFrame1.join(DataFrame2, how = 'left')
       
       
   - 7.6 그룹연산
       - 그룹객체 만들기 : 1개의 열을 기준으로 그룹화 ( 문법 : DataFrame객체.groupby('기준이되는 열')
       - 여러열을 기준으로 그룹화 ( 문법 : DataFrame객체.groupby('기준이되는 열의 리스트')
       - 그룹 연산 메소드( 문법 : DataFrame.max() or .min() or .sum() or .count() or .std() or .var() or .describe() or .info())

.min()  
.sum()  
.count()  
.std()  
.var()  
.describe()  
.info()  

## ■ 시리즈 원소에 함수 매핑

__문법 :__ Series객체.apply('매핑함수')

시리즈 객체에 apply() 메소드를 적용하면 인자로 전달하는 매핑함수에 시리즈의 모든 원소를 하나씩 입력하고 함수의 리턴값을 돌려받는다.

In [163]:
def add_10(n): # 10을 더하는 함수
    return n+10

import seaborn as sns
titanic = sns.load_dataset('titanic')
sr1 = titanic['age'].apply(add_10)
print(sr1)

0      32.0
1      48.0
2      36.0
3      45.0
4      45.0
       ... 
886    37.0
887    29.0
888     NaN
889    36.0
890    42.0
Name: age, Length: 891, dtype: float64


## 문제152. 타이타닉 데이터의 age2라는 파생변수를 생성하는데 기존 나이의 10살을 추가한 파생변수를 생성하시오.

In [164]:
def add_10(n): # 10을 더하는 함수
    return n+10

import seaborn as sns
titanic = sns.load_dataset('titanic')

titanic['age2'] = titanic['age'].apply(add_10)

print(titanic)

     survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alo

## 문제153. (오늘의 마지막 문제) 미국 국민의 의료비 데이터(insurance.csv)를 판다스 데이터 프레임으로 생성하고 나이가 많을 수록 의료비가 많이드는지 회귀분석 모델로 확인하기 위하여 age2라는 파생변수를 생성하고, 기존 나이의 제곱으로 데이터를 생성하시오!

In [169]:
def square(n):
    return n ** 2

import pandas as pd

df = pd.read_csv("d:\\data\\pandas\\insurance.csv")

df['age2'] = df['age'].apply(square)
print(df)

      age     sex     bmi  children smoker     region      charges  age2
0      19  female  27.900         0    yes  southwest  16884.92400   361
1      18    male  33.770         1     no  southeast   1725.55230   324
2      28    male  33.000         3     no  southeast   4449.46200   784
3      33    male  22.705         0     no  northwest  21984.47061  1089
4      32    male  28.880         0     no  northwest   3866.85520  1024
...   ...     ...     ...       ...    ...        ...          ...   ...
1333   50    male  30.970         3     no  northwest  10600.54830  2500
1334   18  female  31.920         0     no  northeast   2205.98080   324
1335   18  female  36.850         0     no  southeast   1629.83350   324
1336   21  female  25.800         0     no  southwest   2007.94500   441
1337   61  female  29.070         0    yes  northwest  29141.36030  3721

[1338 rows x 8 columns]
